onnxruntime/core/providers/cpu/tensor/gather_elements.cc

```diff
@@ -0,0 +1,276 @@
+  // in this context, rank can never be < 1, so saving checking overhead
+  int64_t rank = static_cast<int64_t>(current_dims.size());
+
+  // 'reset' innermost dimension value
+  current_dims[0] = 0;
+
+  // nothing to increment over
+  if (rank == 1) {
+    return;
+  }
+
+  int64_t current_axis = rank - 2;
+
+  while (current_axis >= 0) {
+    if (++current_dims[current_axis] != tensor_dims[current_axis])
+      return;
+
+    current_dims[current_axis] = 0;
+    --current_axis;
+  }
+}
+
+// parse indices_tensor and along the way validate its shape and contents
+std::vector<int64_t> parse_and_validate_indices_tensor(const Tensor* indices_tensor,
+                                                             int64_t axis, const std::vector<int64_t>& input_shape) {
+  const auto& indices_shape = indices_tensor->Shape().GetDims();
+  int64_t indices_rank = static_cast<int64_t>(indices_shape.size());
+  for (int64_t i = 0; i < indices_rank; ++i) {
+    // for all axes except the axis of interest,
+    // make sure that the corresponding 'indices' shape
+    // value if within bounds of the corresponding 'data' shape
+    if (i != axis) {
+      if (indices_shape[i] < 0 || indices_shape[i] > indices_shape[i])
+        ORT_THROW("GatherElements op: 'indices' shape should have values within bounds of 'data' shape");
+    }
+  }
+
+  // first parse 'indices' data
+  auto num_elements = indices_tensor->Shape().Size();
+  std::vector<int64_t> indices_data;
+  // reserving memory ahead as we know the size of the container
+  indices_data.reserve(num_elements);
+  if (indices_tensor->DataType() == DataTypeImpl::GetType<int32_t>()) {
+    const auto* data = indices_tensor->Data<int32_t>();
+    for (int64_t i = 0; i < num_elements; ++i)
+      indices_data.push_back(data[i]);
+  } else if (indices_tensor->DataType() == DataTypeImpl::GetType<int64_t>()) {
+    const auto* data = indices_tensor->Data<int64_t>();
+    for (int64_t i = 0; i < num_elements; ++i)
+      indices_data.push_back(data[i]);
+  } else {
+    ORT_THROW("GatherElements op: Data type for 'indices' tensor must be 'int32_t' and 'int64_t'");
+  }
+
+  // validate 'indices' data
+  // along the way 'fix' negative index values if within bounds
+  int64_t lower_index_limit = -input_shape[axis];
+  int64_t upper_index_limit = input_shape[axis] - 1;
+
+  for (int64_t i = 0; i < num_elements; ++i) {
+    if (indices_data[i] < lower_index_limit || indices_data[i] > upper_index_limit)
+      ORT_THROW("GatherElements op: Value in indices must be within bounds [",
+                lower_index_limit, " , ", upper_index_limit, "]");
+
+    if (indices_data[i] < 0)
+      indices_data[i] += input_shape[axis];
+  }
+
+  return indices_data;
+}
+
+Status GatherElements::CoreImplString(const Tensor* input_tensor, const Tensor* indices_tensor,
+                                      Tensor* output_tensor, int64_t axis) const {
```

Comment:
Do only few lines differ between this can CoreImpl? Could you instead pass a bool for is_string and have input_data and input_data_string, and an if/else for the two places where the copy happens choosing between the string and non-string versions?

Reply:
Well I did think of that, but then that if/else will be within a loop (as we run through the elements in the indices). Seems more performant to dupe some lines of code to avoid checking within the loop.

As for the other optimization (inner dim vs non-inner dim), the checking can occur outside the loop and is easy to put into one method. 

We can still shoe-horn these together by doing NonInnerDimNonString, InnerDimNonString, NonInnerDimString, InnerDimString - but to me this spoils readability of code.

Reply:
Leave it for now. Just feels like a lot of duplication to avoid a branch in an inner loop that should have pretty good prediction. 

another option to consider in the future would be to templatize on the input data type of std::string or int8_t and use remove_pointer and is_class to separate the copylogic. I think the compiler should be able to remove the branch in that case.

Reply:
Templatized on a template parameter - 'is_string' and have a if/else check where the copy happens. Since the check is on a  template parameter, it will be cleaned up by the dead-code optimizer. It will still result in 2 functions when the code is expanded by the pre-processor but still keeps one templatized function in the code-base. I feel this is a better middle-ground.